### DISCUSSION TITLE
Copying placement and routing between SLRs

### DISCUSSION THREAD

**Discussion by bn999** [Xilinx/RapidWright#292] [Q&A] [answered]
I have a design targeting a vu13p which has identical functionality for each of the 4 SLRs.  I have been placing and routing each SLR separately using a Vivado HD flow and piecing them all together at the end because I can get much better timing results than running the entire design at once.  Sometimes one SLR run comes out with particularly good timing.  I would like to reuse it on other SLRs.  Device pins and glue logic (which runs at a much lower clock frequency) are overlayed during the final build stage.

Q: Could a RapidWright script be written to move everything (logic placement, BUFGs, routing) from one SLR's HD .dcp to another SLR?  What would that script look like?

**Comment by eddieh-xlnx** [comment, marked as answer]
Hi @bn999! Great question.

In PR #245 I recently added the `com.xilinx.rapidwright.design.tools.RelocationTools` class which is intended to do what you're asking. It's relatively new and likely not tested in anger, so if there's something missing please report back and I'd be more than happy to help.

**Reply by bn999** [reply]
Thanks!  I'll take a look.

**Reply by bn999** [reply]
All of the methods take offset parameters (e.g. <tile_col_offset> <tile_row_offset>) relating to tile.  But what type of tile are they denominated in?
What would the row_offset be if I wanted to move everything up one SLR?

**Reply by eddieh-xlnx** [reply]
One way to get this information would be to open up Vivado, and look at the height of each SLR in tiles. By clicking on a tile on the top edge and examining Tile Properties, I can see that on a `xcvu13p` its name is `CLE_R_X115Y239` where `Y239` indicates the tile row. Thus, you would want to set this offset to multiples of +/-240 so that everything is moved into the exact same position in a different SLR. Let me know how you get on!

**Reply by bn999** [reply]
So the coordinates are in CLB tile units.
First attempt, it refuses to relocate any RAM or BUFG cells and eventually failed with an error.  I added one line to util/Utils.java and that seems to have fixed things:

```
***************
*** 387,392 ****
--- 387,393 ----
  		sliceDspBramUramTypes.addAll(dspTypes);
  		sliceDspBramUramTypes.addAll(bramTypes);
  		sliceDspBramUramTypes.addAll(uramTypes);
+ 		sliceDspBramUramTypes.addAll(moduleSiteTypes);
  	}
  
  }
```
I suspect that's not the proper solution, but it works.

**Reply by eddieh-xlnx** [reply]
Glad you got it working!

The lines you're referring to are:
https://github.com/Xilinx/RapidWright/blob/a47e670c2b1ff729c4e7d34df5ee6e66b38955c8/src/com/xilinx/rapidwright/util/Utils.java#L385-L389
where `moduleSiteTypes` is:
https://github.com/Xilinx/RapidWright/blob/a47e670c2b1ff729c4e7d34df5ee6e66b38955c8/src/com/xilinx/rapidwright/util/Utils.java#L324-L344
I can see why BUFG cells are not being relocated, but it's not clear why RAM cells are not -- the difference between `bramTypes` and `moduleSiteTypes` appears to be the former is missing:
```
RAMB180,
RAMB181,
FIFO18_0,
FIFO36,
```
and the latter is missing:
```
RAMB18_L,
RAMB18_U
```
which is definitely a bug!

You've also discovered that `RelocationTools` is very conservative in that it only relocates DSP/BRAM/URAM and nothing else:https://github.com/Xilinx/RapidWright/blob/a47e670c2b1ff729c4e7d34df5ee6e66b38955c8/src/com/xilinx/rapidwright/design/tools/RelocationTools.java#L103

Perhaps that should be enlarged to be all `moduleSiteTypes` (though that is far from exhaustive either!)

**Reply by eddieh-xlnx** [reply]
I've gone with a new `RelocationTools.supportedTypes` since it doesn't make as much sense for Laguna sites to be relocated. Proposed changes are in #294.
